Mongo lwc

cryogenian opened the pull request.

cryogenian commented:
[ch3536]

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,143 @@
+      }
+    }
+    override def getArrayAt(cursor: ArrayCursor): BsonValue = cursor match {
+      case (hd :: _) => hd
```
usually we use it without brackets: `case hd :: _ => hd` (below there are some similar usages)

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,143 @@
+      case BsonType.END_OF_DOCUMENT => QNull
+    }
+
+    type ArrayCursor = List[BsonValue]
```
I think this may be a case to use java List here (for performance reasons), so that we don't have to convert after `arr.getValues()` in line 60

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala:
```diff
@@ -0,0 +1,59 @@
+  // calls `new MongoClient` that calls `new ConnectionString` in its internals
+  private def isConnectionString(inp: String): Either[String, String] = {
+    try {
+      val _ = new ConnectionString(inp)
```
Interesting, if I wrap this with `IO` and then catch errors and run it synchronously, would it be effectful? What are effectful criteria?

rintcius replied:
This is side-effecting because it can give different results when the input value is the same. See eg https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/ConnectionString.java#L325 which involves resolving host names

cryogenian replied:
Right. So, allocations or handled exceptions are not effectful?

rintcius replied:
Yeah, handling exceptions is a good way to repair non-total (exception throwing) code.
This implementation would have been ok (with the result type `Either[String, String]`) if `ConnectionString` constructor would only do e.g. a regex check (i.e. if it wasn't side-effecting)

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,101 @@
+    case _ if isRoot(prefixPath) => {
+      mongo.databases.map(_ match {
+        case Database(name) => (ResourceName(name), ResourcePathType.prefix)
+      }).some.pure[F]
```
this can be written like:
```
      mongo.databases
        .map(db => (ResourceName(db.name), ResourcePathType.prefix))
        .some.pure[F]
```

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,101 @@
+
+        dbExists.map(exists => {
+          if (exists) {
+            mongo.collections(db).map(_ match {
```
similar as above, this doesn't need a pattern match

cryogenian commented:
Addressed current comments

+ `.travis.yml` fixes: redundant lines removed, `chmod` removed (I don't really know about `publish`, it's not created on win machine, probably should work), mongo version is set
+ removed matching on `Collection` and `Database`
+ removed `new ConnectionString` thing from `MongoConfig`
+ more java in qdatadecoder
+ Renamed signal -> stream and reorganized imports

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,154 @@
+
+    var entry: Option[Map.Entry[String, BsonValue]] = None
+
+    def hasNext(): Boolean = {
```
This is kinda interesting: the code checks if object or array `hasNext` in qdata and then directly takes values and keys, calling `step` in the last command of while-loop. It's convenient for lists I think, but iterator needs to call `next` at first and then do nothing.

wemrysi replied:
Aye, that's a good point. Maybe instead of `hasNext` this should be named `isEmpty`?

Also, using
```scala
if (iterator.hasNext()) {
  ...
} else {
  false
}
```
is preferable to the try/catch block.

rintcius commented on datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala:
```diff
@@ -0,0 +1,220 @@
+      exists <- mongo.databaseExists(db)
+    } yield exists
+    stream.fold(true)(_ && _).compile.last.unsafeRunSync().getOrElse(false)
+  }
```
Instead of using 1 test in combination with `stream.fold` I'd split these into multiple tests which are easier to check for correctness, like this:

```
    MongoSpec.dbs.foreach { db =>
      s"databaseExists returns true for existing db $db" >> {
        ...
      }
    }
```

(or you can even just take 1 db)

Similar for tests below

cryogenian replied:
Ok

cryogenian commented on .travis.yml:
```diff
@@ -15,26 +17,22 @@ env:
     - set -e
     - "$SBT transferPublishAndTagResources"
-    - TRAVIS_JOB_NUMBER=1 scripts/publishAndTag 'slamdata/quasar-mongo'
+    - TRAVIS_JOB_NUMBER=1 scripts/publishAndTag 'slamdata/quasar-datasource-mongo'
```
Isn't `transferPublishAndTagResources` transfering the script?

rintcius replied:
ah yes, looks like it

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,154 @@
+      case BsonType.JAVASCRIPT_WITH_SCOPE => QString
+      case BsonType.MAX_KEY => QString
+      case BsonType.MIN_KEY => QString
+      case BsonType.END_OF_DOCUMENT => QNull
```
Can we use it? I mean the fact it's regexp

wemrysi replied:
Yeah, this could be a use for `meta`. We did this when encoding to `EJson` where the `meta` value would be an object like `{ "type": "mongodb.regex" }` or similar.

cryogenian commented on datasource/src/test/scala/quasar/physical/mongo/DecodeSpec.scala:
```diff
@@ -0,0 +1,165 @@
+    qdataDecoder.getObjectValueAt(cursor4).isNull() must beTrue
+    val cursor5 = qdataDecoder.stepObject(cursor4)
+    qdataDecoder.hasNextObject(cursor5) must beFalse
+  }
```
I suppose you want me checking objects and arrays, right? Or everything?

rintcius replied:
The different number types would be useful too, they have some tricky cases

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,154 @@
+      case date: BsonDateTime => {
+        Instant.ofEpochMilli(date.getValue()).atOffset(ZoneOffset.UTC)
+      }
+    }
```
ah yes, you are right

cryogenian commented:
+ Db and collection streams refactored to be lists and `Fragment.foreach` is used to create lots of tests instead of one folded

@wemrysi @djspiewak 

We had a discussion with @rintcius about needs of `QDataDecode` property tests. I.e. there is no ability for roundtrip tests, since there is no `QDataEncode` in this lib and it's unclear what exactly to test. 

Also there is another question about `Decimal128` 
+ I think that we should simply convert it to `Real`, because if it's `Decimal128` then db owner probably wants it to be `Decimal128`
+ But probably we can check if it's long or double.
+ Also decimal128 defines `2.0 /= 2.00` and `Real` doesn't preserve this equality rule.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,97 @@
+    } yield res
+  }
+
+  def apply[F[_]: ConcurrentEffect: MonadResourceErr](config: MongoConfig): Stream[F, Mongo[F]] = {
```
For mongo api operations that return a `SingleObservable` (`runCommand` below) I think an `F[_]: Async` would be more convenient than `Stream`. I also think removing `close` from `Mongo` and returning `F[Disposable[F, Mongo[F]]]` here would be safer as users of `Mongo` wouldn't be able to close the client early.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,97 @@
+  def collectionExists(collection: Collection): Stream[F, Boolean] =
+    collections(collection.database).exists(_ === collection)
+
+  def findAll(collection: Collection): Stream[F, BsonValue] = for {
```
I think a return type of `F[Stream[F, BsonValue]]` would be a better fit for the `Datasource` api. The outer `F[_]` represents the existence check and the inner stream the actual data. Additionally, we have `F[_]: MonadResourceErr`,  so prefer `MonadResourceErr.raiseError(...)`.

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,100 @@
+import org.mongodb.scala._
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.api.resource.{ResourceName, ResourcePath}
+import slamdata.Predef._
```
We have actually changed our import style. I think this was talked about in the dev channel, but this was part of why we murdered the wiki reference. The rules are now simply:

- Specific imports, avoiding `_` entirely unless importing implicits or `Predef`
- `slamdata.Predef._` at the top, if relevant
- All imports grouped by top-level package, separated by newlines as relevant
- All groups in alphabetical order, both between themselves and between individual imports
  + Lower-case alphabetizes ahead of upper-case
  + Shorter words alphabetize ahead of longer

I personally prefer a style which is like this, except `scala` and `java` top-level imports get pushed to the end, and I tend to habitually do this by accident, but the above rules are what we have actually pushed for as a team.

wemrysi replied:
I was unaware of this or, at least, had forgotten.

djspiewak replied:
Really? I wonder if we should like, talk about these things more often. Amusingly, this is exactly the import style that you've been following in code you've written recently.

wemrysi replied:
Huh, I pretty much do what @rintcius outlined, so maybe the two aren't very far apart :wink:

djspiewak commented on .travis.yml:
```diff
@@ -15,26 +17,26 @@ env:
     script:
+    - docker run -d -p 127.0.0.1:27018:27017 --name mongodb bitnami/mongodb:latest
     - set -e
+    - sudo chmod +x $SBT
```
Also shouldn't be using `sudo` at all on Travis.

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  def apply[F[_]: ConcurrentEffect: MonadResourceErr](config: MongoConfig): Stream[F, Mongo[F]] = config match {
+    case MongoConfig(conn) => {
```
`s/.flatMap(client => flatMap {/ { client =>/`

cryogenian replied:
@djspiewak what do these `/` mean?

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,97 @@
+
+}
+object Mongo {
+  def observableAsStream[F[_], A](obs: Observable[A])(implicit F: ConcurrentEffect[F]): Stream[F, A] = {
```
A couple issues here

1) Unbounded queues should be avoided so that we can control memory usage, we need to add backpressure via managing calls to `subscription.request(n)` in `onSubscribe` and `onNext`. I'd start with just requesting a single item at a time, we can experiment with batches if performance seems poor.

2) We need to handle early termination of the `Stream` via `subscription.unsubscribe()`. As-is, if a user did a `take(1)` on the returned stream, we'd still stream the entire result, filling up our unbounded queue.

djspiewak replied:
This is super-important and needs to block the merge.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,95 @@
+
+  override def evaluate(path: ResourcePath): F[QueryResult[F]] = {
+    val errorStream =
+      Stream.raiseError(ResourceError.throwableP(ResourceError.pathNotFound(path)))
```
Prefer `MonadResourceErr.raiseError`.

djspiewak commented:
> Also there is another question about Decimal128

We definitely want to get it into the most specific type possible. So if it fits into a `Long` or a `Double`, it should be presented as one of those, because otherwise performance will suck hard. `Real` is basically just an emergency fallback.

> Also decimal128 defines 2.0 /= 2.00 and Real doesn't preserve this equality rule

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,95 @@
+    case _ if isRoot(path) => false.pure[F]
+    case ResourcePath.Leaf(file) => MongoResource.ofFile(file) match {
+      case Some(Database(_)) => false.pure[F]
+      case Some(coll@Collection(_, _)) => mongo.collectionExists(coll).compile.lastOrError
```
I think we want `last.map(_ getOrElse false)` instead of `lastOrError`, the absence of collections shouldn't throw an exception.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,95 @@
+    }
+  }
+  private def isRoot(inp: ResourcePath): Boolean =
+    inp === (ResourcePath.root() / ResourceName("")) || inp === ResourcePath.root()
```
Why are we special casing `ResourcePath.root() / ResourceName("")` here?

cryogenian replied:
`ResourcePath.root() / ResourceName("")` looks like path encoding edge case, that is for some reasone has special case in s3 and azure connectors. If `/` could be parsed as `ResourcePath.root()` and `ResourcePath.root() / ResourceName("")` then it's necessary to handle both as the same thing, otherwise I'm happily remove this strange case out of the code 🙂

wemrysi replied:
Yeah, I'd remove it. If we're getting that as input from somewhere (sdbe?) then we should fix it there instead of special-casing everywhere.

jsantos17 replied:
We probably ought to fix that in `quasar` rather than special casing in all connectors.

rintcius replied:
Yeah same reason as https://github.com/slamdata/quasar-datasource-azure/pull/19

If we remove it here, then SDBE integration won't work at all (repl integration would) so that would mean it needs to be fixed soon. BTW I don't really care if we leave it in for now or remove it, just mentioning it to be explicit.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala:
```diff
@@ -0,0 +1,95 @@
+  val kind: DatasourceType = DatasourceType("mongo", 1L)
+
+  def apply[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](mongo: Mongo[F])
+    : F[MongoDataSource[F]] = ConcurrentEffect[F].delay(new MongoDataSource[F](mongo))
```
Doesn't look like we need to `delay` this, construction seems to be pure.

wemrysi commented:
> We had a discussion with @rintcius about needs of QDataDecode property tests. I.e. there is no ability for roundtrip tests, since there is no QDataEncode in this lib and it's unclear what exactly to test.

I think testing the interesting bits of the decoder is fine. The reason we have the two separate classes is so datasources aren't forced to have a representation for all of `QData` so there isn't any expectation for roundtrip capability.

cryogenian commented:
+ Meta things in decoder
+ Removed weird root case
+ Rewrote `observableAsStream` to use bounded queue and unsubscribe on finish
+ `Mongo.apply` returns `F[Disposable[Mongo, F]]`
+ `findAll` returns `F[Stream[F, Bson]]`

need to 
+ Reorganize imports
+ Fix tests
+ Add a couple of property tests
+ Special handling for `Decimal128`

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -49,49 +55,117 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
   def observableAsStream[F[_], A](obs: Observable[A])(implicit F: ConcurrentEffect[F]): Stream[F, A] = {
-    def handler(cb: Either[Throwable, Option[A]] => Unit): Unit = {
+    @SuppressWarnings(Array("org.wartremover.warts.Var"))
+    def handler(cb: Option[FromObservable[A]] => Unit): Unit = {
```
This still isn't quite going to give us the behavior we need. We _are_ requesting elements one at a time, which is great, but we aren't waiting for the previous item to be processed before requesting the next, which is what is required for backpressure to work.

Instead of requesting the next item in the impl of `onNext` we actually need to capture the effect of requesting the next item and enqueue _that_. See the azure datasource for [an example](https://github.com/slamdata/quasar-datasource-azure/blob/37fe7a239740caa8932a034c8917535df3d18cab/datasource/src/main/scala/quasar/blobstore/azure/rx.scala#L38). Essentially we want a `Queue[F, Either[Throwable, Option[F[A]]]]` where the effect of requesting the next item is captured in `F[_]`.

Lastly, I think an `cats.effect.concurrent.MVar[F, Subscription]` might be a more elegant way to make the `Subscription` accessible outside the `Observer`.

cryogenian commented:
I probably addressed all comments to the pr. 

@rintcius @wemrysi Please, take a look at this again

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,217 @@
+          QNull
+        } else try {
+          val decimal: JDecimal = dec128.bigDecimalValue()
+          if (decimal.scale() < 1 && decimal.compareTo(moreThanMaxLong) < 1 && decimal.compareTo(lessThanMinLong) > -1) {
```
🤦‍♂️

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,172 @@
+    val mkClient: F[MongoClient] =
+      ConcurrentEffect[F].delay(MongoClient(config.connectionString))
+
+    def runCommand(client: MongoClient): F[Unit] = singleObservableAsF[F, Unit](
```
I'll wrap this, but it seems that this is unnecessary, because `client.getDatabase` does nothing that can throw an error

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,172 @@
+  def observableAsStream[F[_]: ConcurrentEffect, A](obs: Observable[A]): Stream[F, A] = {
+    val F = ConcurrentEffect[F]
+    @SuppressWarnings(Array("org.wartremover.warts.Var"))
+    def handler(cb: Option[FromObservable[F[A]]] => Unit): Unit = {
```
Yep, thank you!

cryogenian commented on datasource/src/main/scala/quasar/physical/mongo/decoder.scala:
```diff
@@ -0,0 +1,217 @@
+    @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+    override def getLong(bson: BsonValue): Long = bson match {
+      case bsonDecimal: BsonDecimal128 =>
+        try {
```
Because of `NEGATIVE_ZERO` there is no way to know if it's one of `-0`s from `Decimal128` and it throws `ArithmeticException` during conversion to `BigDecimal`

cryogenian commented on datasource/src/test/scala/quasar/physical/mongo/DecodeSpec.scala:
```diff
@@ -0,0 +1,282 @@
+      refined.asString().getValue() === "minKey"
+    }
+  }
+  "BsonDecimal128 very special cases" >> {
```
:+1:

cryogenian commented:
@wemrysi 

+ Tests for max/min values. 
+ Min/max values fixes
+ suspend for runCommand
+ var -> mvar
+ `FromObservable` deleted
+ indentation fixes

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,143 @@
+    def unsubscribe(s: Subscription): F[Unit] = F.delay {
+      if (!s.isUnsubscribed()) {
+        s.unsubscribe()
+      } else {
```
:golf: The `else` clause can be omitted in this case.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -0,0 +1,143 @@
+      Stream.eval(F.delay(handler(subVar, { x => run(obsQ.enqueue1(x)) })))
+
+    (for {
+      obsQ <- Stream.eval(Queue.boundedNoneTerminated[F, Either[Throwable, F[A]]](32))
```
We can revisit this later if we add batching support, but currently there won't ever be more than 1 item in the queue.

wemrysi commented on datasource/src/test/scala/quasar/physical/mongo/DecodeSpec.scala:
```diff
@@ -0,0 +1,337 @@
+        (BigDecimal(Double.MinValue) * 1.1).bigDecimal
+      ))
+
+    val tinierThanTiny =
```
:laughing:

cryogenian commented:
Thank you very much for review!
